### ISSUE TITLE
Unit tests involving passwordlessWithEmail fail with ExceptionInInitializerError on upgrade to version 2.0.0

### ISSUE THREAD

**Issue by alvindizon** [auth0/Auth0.Android#463]
### Describe the problem

Unit tests involving `AuthenticationAPIClient::passwordlessWithEmail` break on upgrade to version 2.0.0

The full stacktrace is:

```
java.lang.ExceptionInInitializerError
	at com.auth0.android.request.internal.RequestFactory.setupRequest(RequestFactory.kt:82)
	at com.auth0.android.request.internal.RequestFactory.post(RequestFactory.kt:32)
	at com.auth0.android.request.internal.RequestFactory.post(RequestFactory.kt:35)
	at com.auth0.android.authentication.AuthenticationAPIClient.passwordless(AuthenticationAPIClient.kt:531)
	at com.auth0.android.authentication.AuthenticationAPIClient.passwordlessWithEmail(AuthenticationAPIClient.kt:480)
	at com.auth0.android.authentication.AuthenticationAPIClient.passwordlessWithEmail$default(AuthenticationAPIClient.kt:473)
	at com.someapp.login.auth0.AuthManagerImpl$sendEmailCode$1.subscribe(AuthManager.kt:90)
	at io.reactivex.rxjava3.internal.operators.single.SingleCreate.subscribeActual(SingleCreate.java:40)
	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4813)
	at io.reactivex.rxjava3.core.Single.test(Single.java:5513)
	at com.someapp.login.auth0.AuthManagerTest.sendEmailCode returns EmailSent when response is successful(AuthManagerTest.kt:104)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:688)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:210)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:206)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:131)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:65)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:230)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58)
Caused by: java.lang.RuntimeException: Method getMainLooper in android.os.Looper not mocked. See http://g.co/androidstudio/not-mocked for details.
	at android.os.Looper.getMainLooper(Looper.java)
	at com.auth0.android.request.internal.DefaultThreadSwitcher.<init>(ThreadSwitcher.kt:14)
	at com.auth0.android.request.internal.DefaultThreadSwitcher.<clinit>(ThreadSwitcher.kt:13)
```

### What was the expected behavior?
Previously with version 1.30, we have unit tests for testing that a code is sent to an email. We make use of `MockWebServer` that mocks a `200` response from our API, which means that an email has been sent to the specified address. These tests pass before.

### Reproduction

With version 2.0.0, it seems HTTPs usage is enforced, thus we have setup our `MockWebServer` in the `@BeforeEach` part of our test:

        server = MockWebServer().apply {
            useHttps(testSslSocketFactory, false)
            start()
        }

The sslSocketFactory is instantiated lazily:

    /**
     *  Adapted from https://github.com/square/okhttp/blob/master/okhttp-tls/README.md
     */
    private val testSslSocketFactory: SSLSocketFactory by lazy {
        val heldCertificate = HeldCertificate.Builder()
            .commonName("localhost")
            .addSubjectAlternativeName(InetAddress.getByName("localhost").canonicalHostName)
            .build()

        return@lazy HandshakeCertificates.Builder()
            .heldCertificate(heldCertificate)
            .addTrustedCertificate(heldCertificate.certificate)
            .build()
            .sslSocketFactory()
    }

Here is the test itself:

    @Test
    fun `sendEmailCode returns EmailSent when response is successful`() {
        server.enqueue(mockResponse(200, email_success_response))
        authManager.sendEmailCode("me@example.org").test().apply {
            assertValueAt(0){
                (it as EmailResult.EmailSent).email == "me@example.org"
            }
            assertNoToken()
        }
    }

`authManager`'s  `sendEmailCode` uses `AuthenticationAPIClient::passwordlessWithEmail`:

    override fun sendEmailCode(email: String): Single<EmailResult> = Single.create { e ->
        auth0Config.ensureApiClient()?.run {
            passwordlessWithEmail(email, PasswordlessType.ANDROID_LINK).start(object :
                Callback<Void?, AuthenticationException> {
                override fun onSuccess(result: Void?) {
                    e.safeOnSuccess(EmailResult.EmailSent(email))
                }

                override fun onFailure(error: AuthenticationException) {
                    if (error.isNetworkError) {
                        e.safeOnSuccess(EmailResult.NetworkError)
                    } else {
                        e.safeOnSuccess(EmailResult.parseError(error))
                    }
                }
            })
        }
    }

`auth0Config`wraps around the creation of a  `Auth0` object, which in turn is used to create an `AuthenticationAPIClient`. 

    override fun setupAuth0Client() {
        val currentRegion = regionSelectionManager.region
        if (this.region != currentRegion) {
            this.region = currentRegion
            account = when (currentRegion) {
                Region.A -> Auth0(clientIdA, domain)
                Region.B -> Auth0(clientIdB, domain)
            }.apply {
                val client = AuthenticationAPIClient(this)
                apiClient = client
            }
        }
    }

    override fun ensureApiClient(): AuthenticationAPIClient? {
        setupAuth0Client()
        return apiClient
    }

We get a `ExceptionInInitializerError` as described by the full stack trace above.

### Environment

- **Version of this library used:** 2.0.0
- **Which framework are you using, if applicable:** Android
- **Other modules/plugins/libraries that might be involved:** We use JUnit5 and mockk heavily for tests, RxJava 3 for asynchronous operations
- **Any other relevant information you think would be useful:** 

Tried to extend the test with the following rules:

```
@ExtendWith(InstantExecutorExtension::class, RxImmediateScheduler::class)
class AuthManagerTest {
```

But running the tests afterwards results in "Test ignored" in the logs, full stacktrace below:

```
Test ignored.

java.lang.InstantiationException
    at sun.reflect.InstantiationExceptionConstructorAccessorImpl.newInstance(InstantiationExceptionConstructorAccessorImpl.java:48)
    at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
    at org.junit.platform.commons.util.ReflectionUtils.newInstance(ReflectionUtils.java:513)
    at org.junit.platform.commons.util.ReflectionUtils.newInstance(ReflectionUtils.java:488)
    at org.junit.jupiter.engine.extension.MutableExtensionRegistry.registerExtension(MutableExtensionRegistry.java:176)
    at java.util.ArrayList.forEach(ArrayList.java:1257)
    at org.junit.jupiter.engine.extension.MutableExtensionRegistry.createRegistryFrom(MutableExtensionRegistry.java:117)
    at org.junit.jupiter.engine.descriptor.ExtensionUtils.populateNewExtensionRegistryFromExtendWithAnnotation(ExtensionUtils.java:77)
    at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.prepare(ClassBasedTestDescriptor.java:143)
    at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.prepare(ClassBasedTestDescriptor.java:78)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$prepare$1(NodeTestTask.java:111)
    at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.prepare(NodeTestTask.java:111)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:79)
    at java.util.ArrayList.forEach(ArrayList.java:1257)
    at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
    at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
    at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
    at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
    at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
    at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
    at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
    at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
    at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
    at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
    at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
    at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
    at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
    at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
    at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:69)
    at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
    at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:230)
    at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58)
```

**Comment by lbalmaceda** [comment]
This is the key part of the stack trace

```
Caused by: java.lang.RuntimeException: Method getMainLooper in android.os.Looper not mocked.
```

Because [we switch threads](https://github.com/auth0/Auth0.Android/blob/main/auth0/src/main/java/com/auth0/android/request/internal/BaseRequest.kt#L65) internally in the Request class, you need to provide a mock for the looper. We're [currently doing that for the tests](https://github.com/auth0/Auth0.Android/blob/main/auth0/src/test/java/com/auth0/android/authentication/AuthenticationAPIClientTest.kt#L44-L45) using a [Robolectric Shadow](https://github.com/auth0/Auth0.Android/blob/main/auth0/src/test/java/com/auth0/android/request/internal/ThreadSwitcherShadow.java) that runs everything at the moment using their `InlineExecutorService`. You might want to have something similar on your side.

**Comment by alvindizon** [comment]
Hi @lbalmaceda , thanks for your feedback, this is really helpful. We will explore using your suggestion and report back the results.

**Comment by alvindizon** [comment]
Since we're using JUnit5, using Robolectric might be problematic for us now. Looking at the code for `ThreadSwitcher`, it seems it is tied to using `Looper` for posting to main thread. If it's possible in subsequent releases, may I suggest looking at the approach made by Google in their Architecture Components, where they have an `ArchTaskExecutor` class (see code [here](https://androidx.de/androidx/arch/core/executor/ArchTaskExecutor.html)) that centralizes common tasks, and allows users to supply their own task executor. This would allow Junit5 users to create their own Junit5 extension for Auth0, similar to the approach in this [post]([url](https://jeroenmols.com/blog/2019/01/17/livedatajunit5/))

**Comment by lbalmaceda** [comment]
@alvindizon Thanks for the additional context. This is a low priority for us at this moment since Robolectric is required to mock the Android context on the tests and this is an Android-focused SDK. If you can provide a PR with the above suggestion I can review it.

**Comment by alvindizon** [comment]
Hi @lbalmaceda , thanks. Do you have any guidelines before submitting a PR?

**Comment by lbalmaceda** [comment]
This was released in 2.3.0.
Thanks!
